Dockerfile

```diff
@@ -1,15 +1,18 @@
 # cLoki
 FROM node:14-slim
 
-# BUILD FORCE
-ENV BUILD 703030
+# Build ENV
+ENV BUILD 703033
 ENV PORT 3100
 
+# Build App
 COPY . /app
 WORKDIR /app
 RUN npm install
+RUN npm install pm2 -g
```

Comment:
If you want to improve caching and layer difference size between builds when dependencies are not changed you can do something like this:

```suggestion
# Build App
RUN mkdir /app
WORKDIR /app
COPY package.json /app
COPY package-lock.json /app
RUN npm install
RUN npm install pm2 -g
COPY . /app
```

Reply:
Good point. Indeed we intentionally avoided container optimizations so far (see the top ENV forcing rebuilds) due to the high number of changes and library changes in early stages but the time is right. For now I was primarily interested in making sure the switch from directly execution of cloki towards the pm2 execution didn't impact your ability to use the current ENV variables/compose/k8s you are using.

Reply:
> For now I was primarily interested in making sure the switch from directly execution of cloki towards the pm2 execution didn't impact your ability to use the current ENV variables/compose/k8s you are using.

Sure, will check that a bit later

Reply:
@lmangani @derlaft 
`RUN npm install pm2 -g`
should be before 
`COPY package.json /app`
to avoid reinstallation of pm2 each time when packaje.json is changed.
But as far as I know, these dockers are built quite rare and mostly with `--no-cache`, so...

Reply:
>  But as far as I know, these dockers are built quite rare and mostly with --no-cache, so...

1) it's useful for development/testing (I have to personally push test cLoki images to registry via slow DSL)

2) You are using docker build action, pushing directly to registry. This way cache can be exported together with layers ([BuildKit](https://github.com/moby/buildkit#export-cache)) speeding up builds.